Added not fully integrated localization logic, improved Page Navigation.

J0nathan550 opened the pull request:
Fixed: Page Navigation that before was hard coded using string. Problem start to occure when I started developing localization for the Pane strings. I just noticed that when you clicked on buttons (in Pane) and they have different label in them it would not switch page because of the hardcoded Pane string content. It was fixed by adding events.

Added localization logic, the code itself works just fine with no problems. What is left is to bind everywhere the strings with localization we desire. I will continue working on the localization binding. But this is my safe push to keep up with other people. It should cost no problem in code whatsover.

Works by using DynamicResource binding, it grabs from Avalonia Resources specific file of language that you use (default is English) and translates it.

Pages translated: PlayView, MainView (Pane).

What to finish: Fully bind everywhere (where possible strings to translate), create big template English resource file to later add translation for other languages. Implement same localization binding but for the .cs source code (strings inside of source code are not translated, but I hope I will figure out soon how to translate them.).

Video demonstration of this PR:
https://youtu.be/XgD_EiMDU9Y

Plootie commented on SIT.Manager.Avalonia/Models/LocalizationConfig.cs:
```diff
@@ -0,0 +1,55 @@
+        /// Changes translation. Depending on what file was called.
+        /// </summary>
+        /// <param name="targetLanguage">targetLanguage for example just English. no en-US</param>
+        public static void Translate(string targetLanguage)
+        {
+            var translations = App.Current.Resources.MergedDictionaries.OfType<ResourceInclude>().FirstOrDefault(x => x.Source?.OriginalString?.Contains("/Localization/") ?? false);
+
+            if (translations != null) App.Current.Resources.MergedDictionaries.Remove(translations);
+
+            App.Current.Resources.MergedDictionaries.Add(
+                new ResourceInclude(new Uri($"avares://SIT.Manager.Avalonia/Localization/{targetLanguage}.axaml"))
+                {
+                    Source = new Uri($"avares://SIT.Manager.Avalonia/Localization/{targetLanguage}.axaml")
+                });
+        }
+
+        /// <summary>
+        /// Used to show the translated strings when you are starting the application.
+        /// </summary>
+        /// <param name="languageID">number or enum number. to update the translation, useful when you are starting the application and using config loading your language.</param>
+        public static void UpdateTranslationStrings(Languages languageID)
+        {
+            switch (languageID)
+            {
+                case Languages.English:
+                    Translate("English");
+                    break;
+                case Languages.Ukrainian:
+                    Translate("Ukrainian");
+                    break;
+                case Languages.Russian:
+                    Translate("Russian");
+                    break;
+                default:
+                    Translate("English");
+                    break;
+            }
+        }
```
Models should contain minimal to no logic inside of them. This should be pushed into a service

Plootie commented on SIT.Manager.Avalonia/Models/LocalizationConfig.cs:
```diff
@@ -0,0 +1,55 @@
+
+namespace SIT.Manager.Avalonia.Models
+{
+    public static class LocalizationConfig
```
I'm personally not a fan of marking the localization system as static and accessing it globally. This would be better added as a service and injected into classes that need it

Plootie commented on SIT.Manager.Avalonia/Models/LocalizationConfig.cs:
```diff
@@ -0,0 +1,55 @@
+{
+    public static class LocalizationConfig
+    {
+        public enum Languages
```
We should probably use CultureInfo over our own enums

J0nathan550 replied:
> We should probably use CultureInfo over our own enums

Problem behind Culture Info it's not 100% correct when you are starting the application. But still a good point to set at the very first start of the application the language based on what language you have on computer.

Plootie commented on SIT.Manager.Avalonia/ViewModels/MainViewModel.cs:
```diff
@@ -50,6 +51,7 @@ public partial class MainViewModel : ViewModelBase, IRecipient<PageNavigationMes
         ILogger<MainViewModel> logger,
         HttpClient httpClient)
     {
+        UpdateTranslationStrings((Languages)managerConfigService.Config.CurrentLanguageSelected);
```
Accessing the localization system like this gives me the ick, see previous comments about services

Plootie commented on SIT.Manager.Avalonia/Views/MainView.axaml.cs:
```diff
@@ -46,12 +32,53 @@ public partial class MainView : ReactiveUserControl<MainViewModel>
+            return;
         }
     }
-}
+
+    /// <summary>
+    /// <see cref="PlayButton_Clicked(object?, TappedEventArgs)"/> gets fired when you click on PlayButton in Pane View.
+    /// </summary>
+    private void PlayButton_Clicked(object? sender, TappedEventArgs e)
+    {
+        PageNavigation pageNavigation = new(typeof(PlayPage));
+        WeakReferenceMessenger.Default.Send(new PageNavigationMessage(pageNavigation));
+    }
+
+    /// <summary>
+    /// <see cref="ToolsButton_Clicked(object?, TappedEventArgs)"/> gets fired when you click on ToolsButton in Pane View.
+    /// </summary>
+    private void ToolsButton_Clicked(object? sender, TappedEventArgs e)
+    {
+        PageNavigation pageNavigation = new(typeof(ToolsPage));
+        WeakReferenceMessenger.Default.Send(new PageNavigationMessage(pageNavigation));
+    }
+
+    /// <summary>
+    /// <see cref="ServerButton_Clicked(object?, TappedEventArgs)"/> gets fired when you click on ServerButton in Pane View.
+    /// </summary>
+    private void ServerButton_Clicked(object? sender, TappedEventArgs e)
+    {
+        PageNavigation pageNavigation = new(typeof(ServerPage));
+        WeakReferenceMessenger.Default.Send(new PageNavigationMessage(pageNavigation));
+    }
+
+    /// <summary>
+    /// <see cref="ModsButton_Clicked(object?, TappedEventArgs)"/> gets fired when you click on ModsButton in Pane View.
+    /// </summary>
+    private void ModsButton_Clicked(object? sender, TappedEventArgs e)
+    {
+        PageNavigation pageNavigation = new(typeof(ModsPage));
+        WeakReferenceMessenger.Default.Send(new PageNavigationMessage(pageNavigation));
+    }
```
I don't see how this is any better than the current system. In the end it's still hard-coded types and events you're just moving where the hard coded types lie

J0nathan550 replied:
> I don't see how this is any better than the current system. In the end it's still hard-coded types and events you're just moving where the hard coded types lie

Better in they way that you can switch pages that are not depended literally on string. Current system checks what current page name is. And if the page name != whatever in Dictionary we cannot go to another page. Also why to bother? They are meant to be static. I don't think we will add 10.000+ buttons in that pane what so ever.

Plootie replied:
This is actually my fault as I appear to have accidentally swapped from using the NavigationViewItem's Tag property to the InvokedItem which gives the display name. The Tag property allows us to have an identifier separate from the display value without needing an event for each item. I will push the needed changes for this now

Plootie commented on SIT.Manager.Avalonia/Views/SettingsPage.axaml:
```diff
@@ -31,6 +31,11 @@
 								  ToolTip.Tip="Automatically look for updates when opening the launcher."
 								  IsChecked="{Binding Config.LookForUpdates, Mode=TwoWay}"/>
+						<ComboBox Name="languageComboBox" ToolTip.Tip="Translates to your language, select one!" SelectedIndex="{Binding Config.CurrentLanguageSelected, Mode=TwoWay}" SelectionChanged="ComboBox_SelectionChanged">
+							<ComboBoxItem Content="English"/>
+							<ComboBoxItem Content="Ukrainian"/>
+							<ComboBoxItem Content="Russian"/>
```
These item's should be loaded in from a list using data templating for maintainability

Plootie commented on SIT.Manager.Avalonia/Views/SettingsPage.axaml:
```diff
@@ -31,6 +31,11 @@
 						<CheckBox Content="Look For Updates" 
 								  ToolTip.Tip="Automatically look for updates when opening the launcher."
 								  IsChecked="{Binding Config.LookForUpdates, Mode=TwoWay}"/>
+						<ComboBox Name="languageComboBox" ToolTip.Tip="Translates to your language, select one!" SelectedIndex="{Binding Config.CurrentLanguageSelected, Mode=TwoWay}" SelectionChanged="ComboBox_SelectionChanged">
```
I think we should stay away from using an event here. Use IPropertyChanged to notify when the SelectedItem is changed

Plootie commented on SIT.Manager.Avalonia/Views/SettingsPage.axaml.cs:
```diff
@@ -10,5 +11,10 @@ public partial class SettingsPage : UserControl
             this.DataContext = App.Current.Services.GetService<SettingsPageViewModel>();
         }
+
+        private void ComboBox_SelectionChanged(object? sender, SelectionChangedEventArgs e)
+        {
+            UpdateTranslationStrings((Languages)languageComboBox.SelectedIndex);
+        }
```
See previous comment about avoid events

J0nathan550 replied:
I just don't know how to use commands, this thing is complicated for me.

Plootie commented:
I've made a [pull request](https://github.com/stayintarkov/SIT.Manager.Avalonia/pull/34) to fix the NavigationView problem. Once Artehe signs off on it I'll merge it into master for you to update from

J0nathan550 commented:
> I've made a [pull request](https://github.com/stayintarkov/SIT.Manager.Avalonia/pull/34) to fix the NavigationView problem. Once Artehe signs off on it I'll merge it into master for you to update from

I saw that thank you, I will now avoid any events with this one 😄

J0nathan550 commented:
I'm still working, I just fixed issue with my branch. No review required.

J0nathan550 commented:
- Implemented MVVM, Service of Localization.
- Translated almost fully pages.
- Added just logic to be able Translate source code, still I have to translate entire source code. 
Hope you will like it.

Plootie reviewed (changes requested):
I'll hold off reviewing the code for now. This appears to be broken on my end. Selecting any language overwrites the option with "English (United States)" 
![bNwEhVV4f7](https://github.com/stayintarkov/SIT.Manager.Avalonia/assets/34936608/8e58af27-aa16-4571-9bae-515e86d91740)

J0nathan550 commented:
I did one dumb mistake, sorry one second.

Plootie commented:
I'll review this in the morning, There's quite a few things I wanna comment on that'll take too long tonight

Plootie commented on SIT.Manager.Avalonia/Views/MainWindow.axaml:
```diff
@@ -13,8 +13,7 @@
 		WindowStartupLocation="CenterScreen"
-		TransparencyLevelHint="AcrylicBlur"
-		Background="Transparent"
+		Background="Black"
```
Not sure why this is here. It's possible that these options aren't needed but I'm not a fan of changing things unrelated to the PR and undocumented

J0nathan550 replied:
I think it's better to remove AcrylicBlur because we literally do not use transparent AcrylicBlur background, it can make a difference in perfomance when you render a window.

Plootie replied:
Can we move this to another PR. I don't disagree that these are possibly useless but it's outside the scope of this PR and I don't like making undocumented changes

Plootie commented on SIT.Manager.Avalonia/Views/SettingsPage.axaml:
```diff
@@ -22,74 +23,88 @@
+						<CheckBox Content="{DynamicResource SettingsLookForUpdatesTitle}" 
+								  ToolTip.Tip="{DynamicResource SettingsLookForUpdatesTitleToolTip}"
 								  IsChecked="{Binding Config.LookForUpdates, Mode=TwoWay}"/>
+						<Grid>
+							<Grid.ColumnDefinitions>
+								<ColumnDefinition Width="auto"/>
+								<ColumnDefinition/>
+							</Grid.ColumnDefinitions>
```
This can be condense into `<Grid ColumnDefinitions="auto, *">`

J0nathan550 replied:
Sugar syntax but okay 😄

Plootie commented on SIT.Manager.Avalonia/Views/PlayPage.axaml:
```diff
@@ -6,21 +6,22 @@
-			<TextBox Name="UsernameBox" Text="{Binding Username, Mode=TwoWay}" Margin="10 5 10 10" Width="150" HorizontalAlignment="Left" Watermark="Enter Username..."/>
+			<Label Content="{DynamicResource UsernameBoxTitle}" Margin="10 10 10 0" FontFamily="{StaticResource BenderFont}" FontStyle="Oblique"/>
+			<TextBox Name="UsernameBox" Text="{Binding Username, Mode=TwoWay}" Margin="10 5 10 10" HorizontalAlignment="Left" Watermark="{DynamicResource EnterUserNamePlaceholder}"/>
 			<!--Password box-->
-			<Label Content="Password" Margin="10 10 10 0" FontFamily="{StaticResource BenderFont}" FontStyle="Oblique"/>
-			<TextBox Name="PasswordBox" Text="{Binding Password, Mode=TwoWay}" Margin="10 5 10 10" Width="150" HorizontalAlignment="Left" PasswordChar="●" Watermark="Enter Password..."/>
-			<CheckBox Name="RememberMeCheck" Content="Remember Me" Margin="10" IsChecked="{Binding RememberMe}"/>
-			<Button Name="ConnectButton" Command="{Binding ConnectToServerCommand}" Content="Connect" ToolTip.Tip="Connect to the server" Margin="10"/>
-			<Button Name="QuickPlayButton" Command="{Binding QuickPlayCommand}" ToolTip.Tip="Start the Aki server and connect to the server" Margin="10">
+			<Label Content="{DynamicResource PasswordBoxTitle}" Margin="10 10 10 0" FontFamily="{StaticResource BenderFont}" FontStyle="Oblique"/>
+			<TextBox Name="PasswordBox" Text="{Binding Password, Mode=TwoWay}" Margin="10 5 10 10" HorizontalAlignment="Left" PasswordChar="●" Watermark="{DynamicResource EnterPasswordPlaceholder}"/>
+			<CheckBox Name="RememberMeCheck" Content="{DynamicResource RememberMeCheckBoxTitle}" Margin="10" IsChecked="{Binding RememberMe}"/>
+			<Button Name="ConnectButton" Command="{Binding ConnectToServerCommand}" Content="{DynamicResource ConnectButtonTitle}" ToolTip.Tip="{DynamicResource ConnectButtonTitleToolTip}" Margin="10"/>
+			<Button Name="QuickPlayButton" Command="{Binding QuickPlayCommand}" ToolTip.Tip="{DynamicResource StartServerAndConnectButtonTitleToolTip}" Margin="10">
```
Removal of the width attributes here causes the textboxs to shrink to minimum size when you enter a single character as its resizing to largest content and once a character is entered the watermark disappears causing the shrink

J0nathan550 commented on SIT.Manager.Avalonia/Models/LocalizationModel.cs:
But you cannot store everything in CultureInfo, because it's gonna be a waste of memory, better to store few variables. Also it's gonna be for me at least a lot harder to understand and save literally CultureInfo.

J0nathan550 replied:
But I will see, maybe it's not that hard.

Plootie commented on SIT.Manager.Avalonia/Models/ManagerConfig.cs:
```diff
@@ -28,7 +28,7 @@ public partial class ManagerConfig : ObservableObject
         public bool _lookForUpdates = true;
         [ObservableProperty]
-        public LocalizationModel _currentLanguageSelected = new() { FullNameLanguage = CultureInfo.CurrentCulture.NativeName, ShortNameLanguage = CultureInfo.CurrentCulture.Name };
+        public string _currentLanguageSelected = CultureInfo.CurrentCulture.Name;
```
This falls back to the users current culture info, but if the user's culture doesn't match any of the localizations we have the selected field will be blank (In my case this defaults to en-GB, since that doesn't match any localizations we have the selected item becomes blank)

J0nathan550 replied:
Yeah, I forgot about that issue will fix it.

Plootie commented on SIT.Manager.Avalonia/Services/LocalizationService.cs:
```diff
@@ -0,0 +1,119 @@
+        }
+
+        private ResourceInclude? resourceInclude;
+        private string currentLanguage = string.Empty;
+        /// <summary>
+        /// Changes the localization in .cs files that contains strings that you cannot change inside the page.
+        /// Functions contain neat parameters that help modify source strings, like in C#, but inside a Resource file.
+        /// Example will be: Your path is %1. %1 → path. | Output: Your path is: C:\Users\...
+        /// where % is the definition of parameter, and 1…n is the hierarchy of parameters passed to the function.
+        /// </summary>
+        /// <param name="key">string that you are accessing in Localization\*culture-info*.axaml file</param>
+        /// <param name="replaces">parameters in hierarchy, example: %1, %2, %3, "10", "20, "30" | output: 10, 20, 30</param>
+        public string TranslateSource(string key, params string[] replaces)
+        {
+            if (resourceInclude == null || string.IsNullOrEmpty(currentLanguage) || currentLanguage != _configService.Config.CurrentLanguageSelected)
+            {
+                try
+                {
+                    resourceInclude = CreateResourceLocalization($"avares://SIT.Manager.Avalonia/Localization/{_configService.Config.CurrentLanguageSelected}.axaml");
+                }
+                catch // If there was an issue loading current Culture language, we will default by English.
+                {
+                    resourceInclude = CreateResourceLocalization($"avares://SIT.Manager.Avalonia/Localization/en-US.axaml");
+                }
+            }
+
+            string result = "not found";
+            if (resourceInclude.TryGetResource(key, null, out object? translation))
+            {
+                if (translation != null)
+                {
+                    result = (string)translation;
+                    for (int i = 0; i < replaces.Length; i++)
+                    {
+                        result = result.Replace($"%{i + 1}", replaces[i]);
+                    }
+                }
+            }
+            return result;
+        }
```
These two methods have significant overlap, Push some of the duplicate code to another function to avoid repeating the same thing. Also instead of using a try catch were reference the known available localizations to ensure it's valid before attempting

Plootie commented on SIT.Manager.Avalonia/Services/LocalizationService.cs:
```diff
@@ -0,0 +1,119 @@
+        /// </summary>
+        public List<CultureInfo> GetAvailableLocalizations()
+        {
+            List<CultureInfo> result = []; // Initialize the list properly
+            var assembly = typeof(LocalizationService).Assembly;
+            string folderName = string.Format("{0}.Localization", assembly.GetName().Name);
+            result = assembly.GetManifestResourceNames()
+                .Where(r => r.StartsWith(folderName) && r.EndsWith(".axaml"))
+                .Select(r =>
+                {
+                    string[] parts = r.Split('.');
+                    string languageCode = parts[^2];
+                    try
+                    {
+                        CultureInfo cultureInfo = new(languageCode);
+                        return cultureInfo;
+                    }
+                    catch (CultureNotFoundException)
+                    {
+                        CultureInfo cultureInfo = new("en-US");
+                        return cultureInfo;
+                    }
+                })
+                .Where(cultureInfo => cultureInfo != null)
+                .ToList();
+            return result;
+        }
+
+        /// <summary>
+        /// Creates a Resource that will load Localization later on.
+        /// </summary>
+        /// <returns>Resource with Localization</returns>
+        private ResourceInclude CreateResourceLocalization(string url)
+        {
+            var self = new Uri("resm:Styles?assembly=SIT.Manager.Avalonia");
+            return new ResourceInclude(self)
+            {
+                Source = new Uri(url)
+            };
+        }
```
I don't like this approach. We're repeating ourselves massively here. We already pull values such as `SIT.Manager.Avalonia.Localization.en-US.axaml` from the first half of the LINQ expression, only to extract the country code, convert it into a CultureInfo to the reconstruct that exact path again later to create the resource

Plootie commented on SIT.Manager.Avalonia/ViewModels/SettingsPageViewModel.cs:
```diff
@@ -107,7 +122,8 @@ public partial class SettingsPageViewModel : ViewModelBase
     protected override void OnPropertyChanged(PropertyChangedEventArgs e) {
         base.OnPropertyChanged(e);
-
+        _localizationService.Translate(CurrentLocalization);
```
This will call the localization code on every property change regardless of if the localization was actually changed

J0nathan550 replied:
Sure, sorry, I will add propertychange based on if the Settings current localization was changed.

Plootie commented on SIT.Manager.Avalonia/Services/LocalizationService.cs:
My only gripe left with this PR is some of these methods. I'm seeing a lot of `avares://SIT.Manager.Avalonia/Localization/*.axaml` that eventually ends up getting passed down to `CreateResourceLocalization`. This could be changed so that only the file name gets passed in to the method and it constructs the rest of the path itself. Once that's addressed I'm happy to merge this ^^

J0nathan550 replied:
Do you want me to later recreate PR to avoid thousands of commits?

J0nathan550 commented:
Done!

Plootie reviewed (approved):
I think I'm happy with this :3 @artehe Do you mind just giving this a skim to make sure I haven't missed something dumb?

artehe commented on SIT.Manager.Avalonia/ViewModels/SettingsPageViewModel.cs:
```diff
@@ -107,7 +122,11 @@ public partial class SettingsPageViewModel : ViewModelBase
     protected override void OnPropertyChanged(PropertyChangedEventArgs e) {
         base.OnPropertyChanged(e);
-
+        if (e.PropertyName == nameof(CurrentLocalization))
+        {
+            _localizationService.Translate(CurrentLocalization);
+            Config.CurrentLanguageSelected = CurrentLocalization.Name;
+        }
```
I know I said two things but I can't think how to handle the second so we can leave it for now and I'll create an issue to come back to it sometime in the future.  So one last thing then this is good to go. 

Could we take the highlighted lines out and instead use th the specific property change value for CurrentLanguageSelected?

If you wanna just copy and paste the function it'd be this:
```
    partial void OnCurrentLocalizationChanged(CultureInfo value)
    {
        _localizationService.Translate(value);
        Config.CurrentLanguageSelected = value.Name;
    }
```
